sdk/core/azure-core-http-httpurlconnection/src/main/java/com/azure/core/http/httpurlconnection/SocketClient.java

```diff
@@ -0,0 +1,148 @@
+
+    /**
+     * Opens a socket connection, then writes the PATCH request across the
+     * connection and reads the response
+     *
+     * @param httpRequest {@link com.azure.core.http.HttpRequest} instance
+     * @return an instance of HttpUrlConnectionResponse
+     */
+    public static HttpUrlConnectionResponse sendPatchRequest(HttpRequest httpRequest) throws IOException {
+        String host = httpRequest.getUrl().getHost();
+        int port = httpRequest.getUrl().getPort();
+        if (httpRequest.getUrl().getProtocol().equals("https")) {
+            SSLSocketFactory sslSocketFactory = (SSLSocketFactory) SSLSocketFactory.getDefault();
+            try (SSLSocket socket = (SSLSocket) sslSocketFactory.createSocket(host, port)) {
+                return doInputOutput(httpRequest, socket);
+            }
+        } else if (httpRequest.getUrl().getProtocol().equals("http")) {
+            try (Socket socket = new Socket(host, port)) {
+                return doInputOutput(httpRequest, socket);
+            }
+        }
+        throw new ProtocolException("Only HTTP and HTTPS are supported by this client.");
+    }
+
+    /**
+     * Gets a String representation of the request and writes it to the output stream,
+     * then calls buildResponse to get an instance of HttpUrlConnectionResponse from the
+     * input stream
+     *
+     * @param httpRequest {@link com.azure.core.http.HttpRequest} instance
+     * @param socket {@link java.net.Socket} instance
+     * @return an instance of HttpUrlConnectionResponse
+     */
+    private static HttpUrlConnectionResponse doInputOutput(HttpRequest httpRequest, Socket socket) throws IOException {
+        httpRequest.setHeader(HttpHeaderName.HOST, httpRequest.getUrl().getHost());
+        String request = buildPatchRequest(httpRequest);
+        try (BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream(), StandardCharsets.UTF_8));
+             OutputStreamWriter out = new OutputStreamWriter(socket.getOutputStream())) {
+
+            out.write(request);
+            out.flush();
+
+            HttpUrlConnectionResponse response = buildResponse(httpRequest, in);
+
+            String redirectLocation = response.getHeaders().stream()
+                .filter(h -> h.getName().equals("Location"))
+                .map(HttpHeader::getValue)
+                .findFirst()
+                .orElse(null);
+
+            if (redirectLocation != null) {
+                httpRequest.setUrl(redirectLocation);
+                return sendPatchRequest(httpRequest);
+            }
+            return response;
+        }
+    }
+
+    /**
+     * Converts an instance of HttpRequest to a String representation for sending
+     * over the output stream
+     *
+     * @param httpRequest {@link com.azure.core.http.HttpRequest} instance
+     * @return the String representation of the HttpRequest
+     */
+    private static String buildPatchRequest(HttpRequest httpRequest) {
```

Comment:
I think this API would be better off being `void` and taking the stream sending the request. That way it can write the patch request directly to the network instead of buffering it once before sending it.